File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        st = deque()
        n = len(heights)
        ans = -1

        for i, h in enumerate(heights):
            while st and heights[st[-1]] > h:
                prev = -1
                t = st.pop()
                if st:
                    prev = st[-1]
                ans = max(ans, heights[t] * (i - prev - 1))
            
            st.append(i)
        
        while st:
            prev = -1
            t = st.pop()
            if st:
                prev = st[-1]
            ans = max(ans, heights[t] * (n - prev - 1))
        
        return ans

File: 84-largest-rectangle-in-histogram/largest-rectangle-in-histogram.js
/**
 * @param {number[]} heights
 * @return {number}
 */
var largestRectangleArea = function(heights) {
    let st = [];
    let ans = 0;

    for(let i in heights) {
        h = heights[i];
        while(st && heights[st[st.length-1]] > h) {
            let t = st.pop();
            let prev = -1;
            if(st.length > 0) {
                prev = st[st.length-1];
            }
            ans = Math.max(ans, heights[t] * (i - prev - 1));
        }
        st.push(i);
    }
    while(st.length > 0) {
        let t = st.pop();
        let prev = -1;
        if(st.length > 0) {
            prev = st[st.length-1];
        }
        ans = Math.max(ans, heights[t] * (heights.length - prev - 1));
    }
    return ans;
};